From 5cf5c47614d8917320d4cbfed14263ed95de26b9 Mon Sep 17 00:00:00 2001
From: Dinolek <git@dinolek.me>
Date: Thu, 10 Jul 2025 16:55:31 +0200
Subject: [PATCH 089/115] backlight: lm36274_bl: Add ramp time, ovp and
 exponential mode

Add support for configuring ramp time, over voltage protection (OVP)
threshold, and exponential mode in the LM36274 backlight driver.

Signed-off-by: Dinolek <git@dinolek.me>
---
 drivers/video/backlight/lm36274_bl.c | 156 +++++++++++++++++++++++++++
 1 file changed, 156 insertions(+)

diff --git a/drivers/video/backlight/lm36274_bl.c b/drivers/video/backlight/lm36274_bl.c
index 666f568b1..c4174e904 100644
--- a/drivers/video/backlight/lm36274_bl.c
+++ b/drivers/video/backlight/lm36274_bl.c
@@ -17,6 +17,47 @@
 #define LM36274_MAX_STRINGS	4
 #define LM36274_BL_EN		BIT(4)
 
+#define LM36274_OVP_SHIFT	5
+#define LM36274_OVP_MASK	GENMASK(7, LM36274_OVP_SHIFT)
+
+/**
+ * enum @lm36274_ovp_threshold - Over Voltage Protection (OVP) threshold
+ */
+enum lm36274_ovp_threshold {
+	LM36274_OVP_17V = 0,
+	LM36274_OVP_21V = 1,
+	LM36274_OVP_25V = 2,
+	LM36274_OVP_29V = 3,
+};
+
+#define LM36274_RAMP_TIME_SHIFT	3
+#define LM36274_RAMP_TIME_MASK	GENMASK(6, LM36274_RAMP_TIME_SHIFT)
+
+/**
+ * enum @lm36274_ramp_time - Ramp time for brightness change
+ */
+enum lm36274_ramp_time {
+	LM36274_RAMP_TIME_0US = 0,
+	LM36274_RAMP_TIME_500US = 1,
+	LM36274_RAMP_TIME_750US = 2,
+	LM36274_RAMP_TIME_1MS = 3,
+	LM36274_RAMP_TIME_2MS = 4,
+	LM36274_RAMP_TIME_5MS = 5,
+	LM36274_RAMP_TIME_10MS = 6,
+	LM36274_RAMP_TIME_20MS = 7,
+	LM36274_RAMP_TIME_50MS = 8,
+	LM36274_RAMP_TIME_100MS = 9,
+	LM36274_RAMP_TIME_250MS = 10,
+	LM36274_RAMP_TIME_800MS = 11,
+	LM36274_RAMP_TIME_1S = 12,
+	LM36274_RAMP_TIME_2S = 13,
+	LM36274_RAMP_TIME_4S = 14,
+	LM36274_RAMP_TIME_8S = 15,
+};
+
+#define LM36274_BLED_MAP_SHIFT	3
+#define LM36274_BLED_MAP_MASK	GENMASK(3, LM36274_BLED_MAP_SHIFT)
+
 /**
  * struct lm36274
  * @pdev: platform device
@@ -26,6 +67,9 @@
  * @dev: Pointer to the devices device struct
  * @led_sources: The LED strings supported in this array
  * @num_leds: Number of LED strings are supported in this array
+ * @ramp_time: Ramp time for brightness change
+ * @ovp: Over Voltage Protection (OVP) threshold
+ * @exp_mode: Exponential mode for brightness control
  */
 struct lm36274 {
 	struct platform_device *pdev;
@@ -36,6 +80,10 @@ struct lm36274 {
 
 	u32 led_sources[LM36274_MAX_STRINGS];
 	int num_leds;
+
+	enum lm36274_ramp_time ramp_time;
+	enum lm36274_ovp_threshold ovp;
+	bool exp_mode;
 };
 
 static int lm36274_bl_update_status(struct backlight_device *bl_dev)
@@ -54,6 +102,7 @@ static const struct backlight_ops lm36274_bl_ops = {
 static int lm36274_init(struct lm36274 *chip)
 {
 	int enable_val = 0;
+	int ret;
 	int i;
 
 	for (i = 0; i < chip->num_leds; i++)
@@ -66,6 +115,24 @@ static int lm36274_init(struct lm36274 *chip)
 
 	enable_val |= LM36274_BL_EN;
 
+	ret = regmap_write_bits(chip->regmap, LM36274_REG_BL_CFG_1,
+				LM36274_BLED_MAP_MASK,
+				!chip->exp_mode << LM36274_BLED_MAP_SHIFT);
+	if (ret)
+		return ret;
+
+	ret = regmap_write_bits(chip->regmap, LM36274_REG_BL_CFG_1,
+				LM36274_OVP_MASK,
+				chip->ovp << LM36274_OVP_SHIFT);
+	if (ret)
+		return ret;
+
+	ret = regmap_write_bits(chip->regmap, LM36274_REG_BL_CFG_2,
+				LM36274_RAMP_TIME_MASK,
+				chip->ramp_time << LM36274_RAMP_TIME_SHIFT);
+	if (ret)
+		return ret;
+
 	return regmap_write(chip->regmap, LM36274_REG_BL_EN, enable_val);
 }
 
@@ -76,6 +143,8 @@ static int lm36274_parse_dt(struct lm36274 *chip,
 	struct device *dev = chip->dev;
 	struct fwnode_handle *child;
 	u32 brightness;
+	u32 ramp_time;
+	u32 ovp;
 	int ret;
 
 	/* There should only be 1 node */
@@ -115,6 +184,93 @@ static int lm36274_parse_dt(struct lm36274 *chip,
 
 	props->brightness = min_t(u32, brightness, props->max_brightness);
 
+	chip->exp_mode =
+		fwnode_property_read_bool(child, "ti,exponential-mode");
+	props->scale = chip->exp_mode ? BACKLIGHT_SCALE_NON_LINEAR :
+					BACKLIGHT_SCALE_LINEAR;
+
+	ret = fwnode_property_read_u32(child, "ti,ovp-microvolt", &ovp);
+	if (ret) {
+		chip->ovp = LM36274_OVP_21V;
+	} else {
+		switch (ovp) {
+		case 17000000:
+			chip->ovp = LM36274_OVP_17V;
+			break;
+		case 21000000:
+			chip->ovp = LM36274_OVP_21V;
+			break;
+		case 25000000:
+			chip->ovp = LM36274_OVP_25V;
+			break;
+		case 29000000:
+			chip->ovp = LM36274_OVP_29V;
+			break;
+		default:
+			dev_err(dev, "Invalid OVP %d\n", ovp);
+			return -EINVAL;
+		}
+	}
+
+	ret = fwnode_property_read_u32(child, "ti,ramp-time-us", &ramp_time);
+	if (ret) {
+		chip->ramp_time = LM36274_RAMP_TIME_500US;
+	} else {
+		switch (ramp_time) {
+		case 0:
+			chip->ramp_time = LM36274_RAMP_TIME_0US;
+			break;
+		case 500:
+			chip->ramp_time = LM36274_RAMP_TIME_500US;
+			break;
+		case 750:
+			chip->ramp_time = LM36274_RAMP_TIME_750US;
+			break;
+		case 1000:
+			chip->ramp_time = LM36274_RAMP_TIME_1MS;
+			break;
+		case 2000:
+			chip->ramp_time = LM36274_RAMP_TIME_2MS;
+			break;
+		case 5000:
+			chip->ramp_time = LM36274_RAMP_TIME_5MS;
+			break;
+		case 10000:
+			chip->ramp_time = LM36274_RAMP_TIME_10MS;
+			break;
+		case 20000:
+			chip->ramp_time = LM36274_RAMP_TIME_20MS;
+			break;
+		case 50000:
+			chip->ramp_time = LM36274_RAMP_TIME_50MS;
+			break;
+		case 100000:
+			chip->ramp_time = LM36274_RAMP_TIME_100MS;
+			break;
+		case 250000:
+			chip->ramp_time = LM36274_RAMP_TIME_250MS;
+			break;
+		case 800000:
+			chip->ramp_time = LM36274_RAMP_TIME_800MS;
+			break;
+		case 1000000:
+			chip->ramp_time = LM36274_RAMP_TIME_1S;
+			break;
+		case 2000000:
+			chip->ramp_time = LM36274_RAMP_TIME_2S;
+			break;
+		case 4000000:
+			chip->ramp_time = LM36274_RAMP_TIME_4S;
+			break;
+		case 8000000:
+			chip->ramp_time = LM36274_RAMP_TIME_8S;
+			break;
+		default:
+			dev_err(dev, "Invalid ramp time %d\n", ramp_time);
+			return -EINVAL;
+		}
+	}
+
 	return 0;
 err:
 	fwnode_handle_put(child);
-- 
2.56.0

